datasets/flwr_datasets/partitioner/distribution_partitioner.py

```diff
@@ -0,0 +1,336 @@
+            cumsum_division_numbers = np.cumsum(label_samples[unique_label])
+            split_indices = np.split(
+                unique_label_to_indices[unique_label], cumsum_division_numbers
+            )
+            split_indices_per_label[unique_label] = split_indices
+
+        # Initialize sampling tracker. Keys are the unique class labels.
+        # Values are the smallest indices of each array in `label_samples`
+        # which will be sampled next. Once a sample is taken from a label/key,
+        # increment the value (index) by 1.
+        index_tracker = {k: 0 for k in unique_labels}
+
+        # Prepare data structure to store indices assigned to partition ids
+        self._partition_id_to_indices = {
+            partition_id: [] for partition_id in range(self._num_partitions)
+        }
+
+        for partition_id in range(self._num_partitions):
+            # Get the `num_unique_labels_per_partition` labels for each partition. Use
+            # `numpy.roll` to get indices of adjacent sorted labels for pathological
+            # label distributions.
+            labels_per_client = np.roll(unique_labels, -partition_id)[
+                : self._num_unique_labels_per_partition
+            ]
+            for label in labels_per_client:
+                index_to_sample = index_tracker[label]
+                self._partition_id_to_indices[partition_id].extend(
+                    split_indices_per_label[label][index_to_sample]
+                )
+                index_tracker[label] += 1
+
+        # Shuffle the indices to avoid datasets with targets in sequences like
+        # [00000, 11111, ...]) if the shuffle is True
+        if self._shuffle:
+            for indices in self._partition_id_to_indices.values():
+                # In place shuffling
+                self._rng.shuffle(indices)
+        self._partition_id_to_indices_determined = True
+
+    def _check_distribution_array_shape_if_needed(self) -> None:
+        """Test distribution array shape correctness."""
+        if not self._partition_id_to_indices_determined:
+            if not isinstance(self._distribution_array, np.ndarray):
+                raise TypeError("Input must be a numpy array.")
+
+            if self._distribution_array.ndim != 2:
+                raise ValueError("The distribution array is not 2-dimensional.")
+
+            self._num_unique_labels = len(self.dataset.unique(self._partition_by))
+            num_columns = (
+                self._num_unique_labels_per_partition
+                * self._num_partitions
+                / self._num_unique_labels
+            )
+            if self._distribution_array.shape != (self._num_unique_labels, num_columns):
+                raise ValueError(
+                    f"The distribution array shape is {self._distribution_array.shape},"
+                    f" but it should be ({self._num_unique_labels}, {num_columns})."
```

Comment:
Let's make it more obvious what is the error problem e.g. if the first dim value is not properly set let's explain that the number of unique labels in the dataset is `x` but given `y`. And for the second shape mismatch, e.g., we can defer to the documentation or note that it might be caused by the first shape mismatch in case both are wrong.